test/cpp/api/moduledict.cpp

```diff
@@ -0,0 +1,304 @@
+    {"conv", Conv2d(1, 2, 3).ptr()},
+    {"dropout", Dropout(0.5).ptr()},
+  };
+  ModuleDict dict(orderdict);
+  const auto& keys = dict->keys();
+  std::vector<std::string> expected{"linear", "conv", "dropout"};
+  ASSERT_EQ(keys, expected);
+  ASSERT_THROWS_WITH(dict["batch"], " 'batch' is not defined");
+
+  ASSERT_TRUE(dict["linear"]->as<Linear>());
+  ASSERT_TRUE(dict["conv"]->as<Conv2d>());
+  ASSERT_TRUE(dict["dropout"]->as<Dropout>());
+}
+
+TEST_F(ModuleDictTest, Values) {
+  struct M : Module {
+    explicit M(int value_) : value(value_) {}
+    int value;
+  };
+
+  torch::OrderedDict<std::string, std::shared_ptr<Module>> orderdict = {
+    {"module_1", std::make_shared<M>(1)},
+    {"module_2", std::make_shared<M>(2)},
+  };
+  ModuleDict dict(orderdict);
+  const auto& values = dict->values();
+  const auto& expected = orderdict.values();
+  ASSERT_EQ(values, expected);
+  ASSERT_TRUE(std::equal(
+      dict->begin(),
+      dict->end(),
+      orderdict.begin(),
+      [](const auto& lhs,
+         const auto& rhs) {
+        return lhs.value().get() == rhs.value().get();
+      }));
+}
+
+TEST_F(ModuleDictTest, SanityCheckForHoldingStandardModules) {
+  torch::OrderedDict<std::string, std::shared_ptr<Module>> orderdict = {
+    {"linear", Linear(10, 3).ptr()},
+    {"conv", Conv2d(1, 2, 3).ptr()},
+    {"dropout", Dropout(0.5).ptr()},
+    {"batch", BatchNorm2d(5).ptr()},
+    {"embedding", Embedding(4, 10).ptr()},
+    {"lstm", LSTM(4, 5).ptr()}
+  };
+  ModuleDict dict(orderdict);
+}
+
+TEST_F(ModuleDictTest, HasReferenceSemantics) {
+  torch::OrderedDict<std::string, std::shared_ptr<Module>> orderdict = {
+    {"linear1", Linear(2, 3).ptr()},
+    {"linear2", Linear(3, 4).ptr()},
+    {"linear3", Linear(4, 5).ptr()},
+  };
+  ModuleDict first(orderdict);
+  ModuleDict second(orderdict);
+
+  ASSERT_EQ(first->size(), second->size());
+  ASSERT_TRUE(std::equal(
+      first->begin(),
+      first->end(),
+      second->begin(),
+      [](const auto& lhs,
+         const auto& rhs) {
+        return lhs.value().get() == rhs.value().get();
+      }));
+}
+
+TEST_F(ModuleDictTest, IsCloneable) {
+  torch::OrderedDict<std::string, std::shared_ptr<Module>> orderdict = {
+    {"linear", Linear(2, 3).ptr()},
+    {"relu", Functional(torch::relu).ptr()},
+    {"batch", BatchNorm1d(3).ptr()},
+  };
+  ModuleDict dict(orderdict);
+  ModuleDict clone = std::dynamic_pointer_cast<ModuleDictImpl>(dict->clone());
+  ASSERT_EQ(dict->size(), clone->size());
+
+  for (auto it = dict->begin(), it_c = clone->begin(); it != dict->end(); ++it, ++it_c) {
+    // The key should be same
+    ASSERT_EQ(it->key(), it_c->key());
+    // The modules should be the same kind (type).
+    ASSERT_EQ(it->value()->name(), it_c->value()->name());
+    // But not pointer-equal (distinct objects).
+    ASSERT_NE(it->value(), it_c->value());
+  }
+
+  // Verify that the clone is deep, i.e. parameters of modules are cloned too.
+  torch::NoGradGuard no_grad;
```

Comment:
Why is this needed?

Reply:
Yes, no need to run autograd for the following function as the test is to verify clone..